gplugins/lumerical/eme.py

```diff
@@ -0,0 +1,1063 @@
+
+        if run_mode_convergence:
+            if not hide:
+                logger.info("Running mode convergence.")
+            self.mode_convergence_data = self.update_mode_convergence(plot=not hide)
+
+        if not hide:
+            plt.show()
+
+    def update_mesh_convergence(self, plot: bool = False) -> pd.DataFrame:
+        """
+        Update simulation based on mesh convergence testing. Updates both Lumerical session and simulation settings.
+
+        Parameters:
+            plot: Plot and save convergence results
+
+        Returns:
+            Convergence results
+        """
+
+        s = self.session
+        cs = self.convergence_settings
+        ss = self.simulation_settings
+
+        s21 = []
+        s11 = []
+        mesh_cells_per_wavl = []
+        converged = False
+        while not converged:
+            s.switchtolayout()
+            s.set("dy", ss.wavelength / ss.mesh_cells_per_wavelength * um)
+            s.set("dz", ss.wavelength / ss.mesh_cells_per_wavelength * um)
+            # Get sparams and refine mesh
+            s.run()
+            s.emepropagate()
+            S = s.getresult("EME", "user s matrix")
+            s11.append(abs(S[0, 0]) ** 2)
+            s21.append(abs(S[1, 0]) ** 2)
+            mesh_cells_per_wavl.append(ss.mesh_cells_per_wavelength)
+
+            ss.mesh_cells_per_wavelength += 1
+
+            # Check whether convergence has been reached
+            if len(s21) > cs.passes or len(s11) > cs.passes:
+                # Calculate maximum diff in sparams
+                sparam_diff = max(
+                    [
+                        max(abs(s21[-1] - np.array(s21[-(cs.passes + 1) : -1]))),
+                        max(abs(s11[-1] - np.array(s11[-(cs.passes + 1) : -1]))),
+                    ]
+                )
+                if plot:
+                    logger.info(
+                        f"Mesh cells per wavelength: {ss.mesh_cells_per_wavelength} | S-Param Diff: {sparam_diff}"
+                    )
+                if sparam_diff < cs.sparam_diff:
+                    converged = True
+                else:
+                    converged = False
+
+        if plot:
+            plt.figure()
+            plt.plot(mesh_cells_per_wavl, s21)
+            plt.plot(mesh_cells_per_wavl, s11)
+            plt.legend(["|S21|^2", "|S11|^2"])
+            plt.grid("on")
+            plt.xlabel("Mesh Cells Per Wavelength")
+            plt.ylabel("Magnitude")
+            plt.title(f"Mesh Convergence | Wavelength={ss.wavelength}um")
+            plt.savefig(
+                str(
+                    self.dirpath
+                    / f"{self.component.name}_convergence"
+                    / f"{self.component.name}_eme_mesh_convergence.png"
+                )
+            )
+
+        convergence_data = pd.DataFrame.from_dict(
+            {"mesh_cells": mesh_cells_per_wavl, "s21": list(s21), "s11": list(s11)}
+        )
+        convergence_data.to_csv(
```

Comment:
maybe we should have a `save()` method that takes an enum or something for the kind of data? thinking about how we can make this as easy as possible to map to a database (rather than individual files) later